zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt

```diff
@@ -83,25 +84,22 @@ class Endpoint internal constructor(
     ) {
       val handler = takeHandler(instanceName, funName)
       val cancelCallbackName = cancelCallbackName(suspendCallbackName)
+      val cancelCallback = RealCancelCallback()
+      bind<CancelCallback>(cancelCallbackName, cancelCallback)
       scope.launch {
-        bind<CancelCallback>(cancelCallbackName, object : CancelCallback {
-          override fun cancel() {
-            this@launch.cancel()
-          }
-        })
-        try {
-          val suspendCallback = take<SuspendCallback>(suspendCallbackName)
-          val inboundCall = InboundCall(handler.context, funName, encodedArguments)
-          val result = try {
-            handler.callSuspending(inboundCall)
-          } catch (e: Exception) {
-            inboundCall.resultException(e)
-          }
-          scope.ensureActive() // Don't resume a continuation if the Zipline has since been closed.
-          suspendCallback.call(result)
-        } finally {
+        cancelCallback.initScope(this@launch)
+        coroutineContext.job.invokeOnCompletion {
           remove(cancelCallbackName)
         }
```

Comment:
The above suggestions (especially the second one) would also simplify this allowing it to exist outside the launched block. This feels like a weird pattern in generally since I think a try/finally would work equally as well.